pkg/config/release.go

```diff
@@ -163,6 +183,340 @@ func GetChangedTemplates(path, baseRev string) ([]ConfigMapSource, error) {
+					return err
+				}
+				change.Reference = &step
+			} else if strings.HasSuffix(filename, "chain.yaml") {
+				step := api.RegistryChainConfig{}
+				err := yaml.Unmarshal(bytes, &step)
+				if err != nil {
+					return err
+				}
+				change.Chain = &step
+			} else if strings.HasSuffix(filename, "workflow.yaml") {
+				step := api.RegistryWorkflowConfig{}
+				err := yaml.Unmarshal(bytes, &step)
+				if err != nil {
+					return err
+				}
+				change.Workflow = &step
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		return RegistryChanges{}, err
+	}
+	if change == nil {
+		return RegistryChanges{}, fmt.Errorf("could not find component")
+	}
+	return *change, nil
+}
+
+// GetChangedRegistrySteps identifies all registry components (refs, chains, and workflows) that changed.
+// After all changes have been identified, the list gets "trimmed": all changed refs that are included in
+// changed refs are removed from the list since the refs will be tested when the chain gets tested. This
+// is done for changed chains included in changed workflows. This helps reduce the number of rehearsals
+// that are performed.
+func GetChangedRegistrySteps(path, baseRev string, refs map[string]api.LiteralTestStep, chains map[string][]api.TestStep, workflows map[string]api.MultiStageTestConfiguration) ([]RegistryChanges, error) {
+	revChanges, err := getRevChanges(path, RegistryPath, baseRev, true)
+	if err != nil {
+		return nil, err
+	}
+	var registryChanges []RegistryChanges
+	if err != nil {
+		return nil, fmt.Errorf("Failed to load registry: %v", err)
+	}
+	fullRegistry := registry{refs: refs, chains: chains, workflows: workflows}
+	cToS := generateCommandToSteps(fullRegistry)
+	seenReference := make(map[string]bool)
+	for _, c := range revChanges {
+		if filepath.Ext(c.Filename) == ".yaml" {
+			change, err := loadRegistryChange(filepath.Join(path, RegistryPath), c.Filename)
+			if err != nil {
+				return nil, err
+			}
+			if change.Reference != nil {
+				if _, ok := seenReference[change.Reference.Reference.As]; ok {
+					continue
+				}
+				seenReference[change.Reference.Reference.As] = true
+			}
+			registryChanges = append(registryChanges, change)
+		} else if strings.HasSuffix(c.Filename, "commands.sh") {
+			stepFilename, err := getStepsFromCommands(c.Filename, c.SHA, cToS)
+			if err != nil {
+				logrus.Errorf("%v", err)
+			}
+			change, err := loadRegistryChange(filepath.Join(path, RegistryPath), stepFilename)
+			if err != nil {
+				return nil, err
+			}
+			if change.Reference != nil {
+				if _, ok := seenReference[change.Reference.Reference.As]; ok {
+					continue
+				}
+				seenReference[change.Reference.Reference.As] = true
+			}
+			registryChanges = append(registryChanges, change)
+		}
+	}
+	trimmedChanges := trimRegistryChanges(registryChanges, fullRegistry)
+	return trimmedChanges, nil
+}
+
+// unrollChains expands/unrolls all chains recursively into references.
+// Borrowed from pkg/resolver/resolver.go
```

Comment:
why does it need to be borrowed?

Reply:
Currently, there is an `unrollChains` function that is almost identical in `pkg/registry`, but it is not exported and it is a function of the internal registry type in `pkg/registry`. Looking at the function, it doesn't actually modify the internal `registry` type that it operates on and it only needs the map of chains, so it could be easily exported.

I'll export that function and make the modifications needed so it doesn't need to be duplicated.